File: lab3/zad2/go/stack/stack_test.go
package stack

import (
	"crypto/rand"
	"math/big"
	"os"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
	"github.com/stretchr/testify/suite"
)

const alphanum = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"

func generateRandomString(length int) string {
	b := make([]byte, length)
	for i := range b {
		n, _ := rand.Int(rand.Reader, big.NewInt(int64(len(alphanum))))
		b[i] = alphanum[n.Int64()]
	}
	return string(b)
}

type StackTestSuite struct {
	suite.Suite
}

func TestStackTestSuite(t *testing.T) {
	suite.Run(t, new(StackTestSuite))
}

func (s *StackTestSuite) TestConstructorDestructor() {
	stack1 := NewStack()
	assert.NotNil(s.T(), stack1)
	assert.NotPanics(s.T(), func() {
		stack1.Print()
	})

	stack2 := NewStack()
	assert.NotNil(s.T(), stack2)
}

func (s *StackTestSuite) TestPushPop() {
	stack := NewStack()
	str1 := generateRandomString(10)
	str2 := generateRandomString(10)
	str3 := generateRandomString(10)

	assert.NotPanics(s.T(), func() {
		stack.Push(str1)
	})
	assert.NotPanics(s.T(), func() {
		stack.Push(str2)
	})
	assert.NotPanics(s.T(), func() {
		stack.Push(str3)
	})

	assert.NotPanics(s.T(), func() {
		stack.Print()
	})

	err := stack.Pop()
	assert.NoError(s.T(), err)
	err = stack.Pop()
	assert.NoError(s.T(), err)
	err = stack.Pop()
	assert.NoError(s.T(), err)

	err = stack.Pop()
	assert.Error(s.T(), err)
}

func (s *StackTestSuite) TestPrintEmpty() {
	stack := NewStack()
	assert.NotPanics(s.T(), func() {
		stack.Print()
	})
}

func (s *StackTestSuite) TestFileOperations() {
	filename := "test_stack.txt"
	stack1 := NewStack()
	stack2 := NewStack()

	testData := make([]string, 5)
	for i := 0; i < 5; i++ {
		str := generateRandomString(10)
		testData[i] = str
		stack1.Push(str)
	}

	outFile, err := os.Create(filename)
	require.NoError(s.T(), err)
	err = stack1.SaveToFile(outFile)
	assert.NoError(s.T(), err)
	outFile.Close()

	inFile, err := os.Open(filename)
	require.NoError(s.T(), err)
	err = stack2.LoadFromFile(inFile)
	assert.NoError(s.T(), err)
	inFile.Close()

	assert.NotPanics(s.T(), func() {
		stack1.Print()
	})
	assert.NotPanics(s.T(), func() {
		stack2.Print()
	})

	os.Remove(filename)
}

func (s *StackTestSuite) TestBinaryFileOperations() {
	filename := "test_stack_binary.bin"
	stack1 := NewStack()
	stack2 := NewStack()

	testData := make([]string, 5)
	for i := 0; i < 5; i++ {
		str := generateRandomString(10)
		testData[i] = str
		stack1.Push(str)
	}

	outFile, err := os.Create(filename)
	require.NoError(s.T(), err)
	err = stack1.SaveToBinaryFile(outFile)
	assert.NoError(s.T(), err)
	outFile.Close()

	inFile, err := os.Open(filename)
	require.NoError(s.T(), err)
	err = stack2.LoadFromBinaryFile(inFile)
	assert.NoError(s.T(), err)
	inFile.Close()

	assert.NotPanics(s.T(), func() {
		stack1.Print()
	})
	assert.NotPanics(s.T(), func() {
		stack2.Print()
	})

	os.Remove(filename)
}

func (s *StackTestSuite) TestExceptionSafety() {
	stack := NewStack()

	err := stack.Pop()
	assert.Error(s.T(), err)

	testStr := generateRandomString(10)
	stack.Push(testStr)
	err = stack.Pop()
	assert.NoError(s.T(), err)
	err = stack.Pop()
	assert.Error(s.T(), err)
}

func (s *StackTestSuite) TestFileOperationsErrors() {
	stack := NewStack()

	err := stack.SaveToFile(nil)
	assert.Error(s.T(), err)
	assert.Contains(s.T(), err.Error(), "file is nil")

	err = stack.LoadFromFile(nil)
	assert.Error(s.T(), err)
	assert.Contains(s.T(), err.Error(), "file is nil")

	err = stack.SaveToBinaryFile(nil)
	assert.Error(s.T(), err)
	assert.Contains(s.T(), err.Error(), "file is nil")

	err = stack.LoadFromBinaryFile(nil)
	assert.Error(s.T(), err)
	assert.Contains(s.T(), err.Error(), "file is nil")

	filename := "test_nonexistent.txt"
	file, err := os.Open(filename)
	require.Error(s.T(), err)
	if file != nil {
		err = stack.LoadFromFile(file)
		assert.Error(s.T(), err)
		file.Close()
	}

	filename = "test_corrupted.bin"
	corruptedData := []byte{0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00}
	err = os.WriteFile(filename, corruptedData, 0644)
	require.NoError(s.T(), err)

	inFile, err := os.Open(filename)
	require.NoError(s.T(), err)
	err = stack.LoadFromBinaryFile(inFile)
	assert.Error(s.T(), err)
	inFile.Close()

	os.Remove(filename)
}

File: lab3/zad2/cpp/stack/stacktest.cpp
#define BOOST_TEST_MODULE StackTests
#define BOOST_TEST_DYN_LINK
#include <boost/test/unit_test.hpp>
#include <random>
#include <algorithm>
#include "../../../zad1/cpp/stack.h"

using namespace std;

string generateRandomString(size_t length = 10) {
    static const char alphanum[] =
        "0123456789"
        "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
        "abcdefghijklmnopqrstuvwxyz";
    
    static mt19937 rng(random_device{}());
    uniform_int_distribution<size_t> dist(0, sizeof(alphanum) - 2);
    
    string result;
    result.reserve(length);
    for (size_t i = 0; i < length; ++i) {
        result += alphanum[dist(rng)];
    }
    return result;
}

BOOST_AUTO_TEST_SUITE(StackTests)

BOOST_AUTO_TEST_CASE(TestConstructorDestructor) {
    Stack stack1;
    BOOST_CHECK_NO_THROW(stack1.print());
    
    Stack* stack2 = new Stack();
    BOOST_CHECK_NO_THROW(delete stack2);
}

BOOST_AUTO_TEST_CASE(TestPushPop) {
    Stack stack;
    string str1 = generateRandomString();
    string str2 = generateRandomString();
    string str3 = generateRandomString();
    
    BOOST_CHECK_NO_THROW(stack.push(str1));
    BOOST_CHECK_NO_THROW(stack.push(str2));
    BOOST_CHECK_NO_THROW(stack.push(str3));
    
    BOOST_CHECK_NO_THROW(stack.print());
    
    BOOST_CHECK_NO_THROW(stack.pop());
    BOOST_CHECK_NO_THROW(stack.pop());
    BOOST_CHECK_NO_THROW(stack.pop());
    
    BOOST_CHECK_THROW(stack.pop(), runtime_error);
}

BOOST_AUTO_TEST_CASE(TestPrintEmpty) {
    Stack stack;
    BOOST_CHECK_NO_THROW(stack.print());
}

BOOST_AUTO_TEST_CASE(TestFileOperations) {
    const string filename = "test_stack.txt";
    Stack stack1, stack2;
    
    vector<string> testData;
    for (int i = 0; i < 5; ++i) {
        string str = generateRandomString();
        testData.push_back(str);
        stack1.push(str);
    }
    
    ofstream outFile(filename);
    BOOST_REQUIRE(outFile.is_open());
    stack1.saveToFile(outFile);
    outFile.close();
    
    ifstream inFile(filename);
    BOOST_REQUIRE(inFile.is_open());
    stack2.loadFromFile(inFile);
    inFile.close();
    
    BOOST_CHECK_NO_THROW(stack1.print());
    BOOST_CHECK_NO_THROW(stack2.print());
    
    remove(filename.c_str());
}

BOOST_AUTO_TEST_CASE(TestBinaryFileOperations) {
    const string filename = "test_stack_binary.bin";
    Stack stack1, stack2;
    
    vector<string> testData;
    for (int i = 0; i < 5; ++i) {
        string str = generateRandomString();
        testData.push_back(str);
        stack1.push(str);
    }
    
    ofstream outFile(filename, ios::binary);
    BOOST_REQUIRE(outFile.is_open());
    stack1.saveToBinaryFile(outFile);
    outFile.close();
    
    ifstream inFile(filename, ios::binary);
    BOOST_REQUIRE(inFile.is_open());
    stack2.loadFromBinaryFile(inFile);
    inFile.close();
    
    BOOST_CHECK_NO_THROW(stack1.print());
    BOOST_CHECK_NO_THROW(stack2.print());
    
    remove(filename.c_str());
}

BOOST_AUTO_TEST_CASE(TestExceptionSafety) {
    Stack stack;
    
    BOOST_CHECK_THROW(stack.pop(), runtime_error);
    
    string testStr = generateRandomString();
    stack.push(testStr);
    BOOST_CHECK_NO_THROW(stack.pop());
    BOOST_CHECK_THROW(stack.pop(), runtime_error);
}


BOOST_AUTO_TEST_SUITE_END()